### PR TITLE
Fix path_watcher

### PR DESCRIPTION
**Pull request by cirospaciari**
### What does this PR do?

- [ ] Documentation or TypeScript types (it's okay to leave the rest blank in this case)
- [x] Code changes

### How did you verify your code works?

Existing tests

### CODE REVIEW

**Review by Jarred-Sumner — CHANGES_REQUESTED** [review]
This does not fix it? The `finalized` field is set to true in the same scope  as the struct is destroyed, which means all cases the struct  is used after finalized is a use after free. 

https://github.com/oven-sh/bun/blob/044a6facdb3dd624b65960fd390de24980df4989/src/bun.js/node/path_watcher.zig#L816

**Comment by cirospaciari** [comment]
> finalized

It's only used when `pending_directories > 0` (on another thread) finalized just say that the owner is gone, when the last task ends/cancel it will call deinit again (but not emit/flush).

**Comment by Jarred-Sumner** [comment]
It's not threadsafe though. The finalized boolean would need to be loaded and stored atomically with a @fence to ensure memory ordering. But even that's not safe because the other thread could be accessing it after deinit

**Comment by cirospaciari** [comment]
> It's not threadsafe though. The finalized boolean would need to be loaded and stored atomically with a @Fence to ensure memory ordering. But even that's not safe because the other thread could be accessing it after deinit

I used mutex to guarantee access to `finalized` and `pending_directories` but I found another mistake let me finish it, and you are right about memory ordering

**Comment by Jarred-Sumner** [comment]
Not all usages are locked by a mutex. If it were, it would hang until after it’s freed and then segfault

File: src/bun.js/node/path_watcher.zig [hunk 1]
```diff
@@ -431,19 +458,20 @@ pub const PathWatcherManager = struct {
...
-                    watcher.mutex.unlock();
-                    manager._decrementPathRef(entry_path_z);
-                    return err;
-                };
-                watcher.mutex.unlock();
+                {
+                    watcher.mutex.lock();
+                    defer watcher.mutex.unlock();
+                    watcher.file_paths.push(bun.default_allocator, child_path.path) catch |err| {
+                        manager._decrementPathRef(entry_path_z);
```
Comment (Jarred-Sumner) [inline src/bun.js/node/path_watcher.zig:465]: dead lock

File: src/bun.js/node/path_watcher.zig [hunk 2]
```diff
@@ -338,46 +365,57 @@ pub const PathWatcherManager = struct {
...
+            var routine: *DirectoryRegisterTask = undefined;
+            {
+                manager.mutex.lock();
+                defer manager.mutex.unlock();
+
+                // use the same thread for the same fd to avoid race conditions
+                if (manager.current_fd_task.getEntry(path.fd)) |entry| {
+                    routine = entry.value_ptr.*;
+
+                    if (watcher.refPendingDirectory()) {
```
Comment (Jarred-Sumner) [inline src/bun.js/node/path_watcher.zig:380]: dead lock

File: src/bun.js/node/path_watcher.zig [hunk 3]
```diff
@@ -338,46 +365,57 @@ pub const PathWatcherManager = struct {
...
+                routine.* = DirectoryRegisterTask{
+                    .manager = manager,
+                    .path = path,
+                    .watcher_list = bun.BabyList(*PathWatcher).initCapacity(bun.default_allocator, 1) catch |err| {
+                        bun.default_allocator.destroy(routine);
+                        return err;
+                    },
+                };
+                errdefer routine.deinit();
+                if (watcher.refPendingDirectory()) {
```
Comment (Jarred-Sumner) [inline src/bun.js/node/path_watcher.zig:401]: dead lock

File: src/bun.js/node/path_watcher.zig [hunk 4]
```diff
@@ -338,46 +365,57 @@ pub const PathWatcherManager = struct {
...
+                    };
+                } else {
+                    return error.UnexpectedFailure;
+                }
+                manager.current_fd_task.put(path.fd, routine) catch |err| {
+                    watcher.unrefPendingDirectory();
                     return err;
                 };
+            }
+            if (manager.refPendingTask()) {
```
Comment (Jarred-Sumner) [inline src/bun.js/node/path_watcher.zig:414]: dead lock

File: src/bun.js/node/path_watcher.zig (hunk 3 above)
Comment (cirospaciari) [reply src/bun.js/node/path_watcher.zig:401]: manager uses a different lock from watcher

File: src/bun.js/node/path_watcher.zig (hunk 4 above)
Comment (cirospaciari) [reply src/bun.js/node/path_watcher.zig:414]: unlock with defer after last context exits

**Comment by Jarred-Sumner** [comment]
Thank you

**Pull request merged**